unhandled item type: 0x86

I send out ListIdentity command as broadcast to scan a network for devices. Usually every Ethernet/IP device is responding as expected. One customer has a device in its network that sends a response which can't be decoded.
When I call EnipPacket.decode(byteBuf) then I get The following exception:
java.io.IOException: java.util.concurrent.ExecutionException: io.netty.handler.codec.DecoderException: unhandled item type: 0x86
	at xxxx.scanEthernetIp(xxxx.java:526)
	at xxxx.scanNetwork(xxxx.java:347)
	at xxxx.scanNetwork(xxxx.java:240)
	at xxxx.lambda$scanNetwork$1(xxxx.java:734)
	at java.lang.Thread.run(Thread.java:748)
Caused by: java.util.concurrent.ExecutionException: io.netty.handler.codec.DecoderException: unhandled item type: 0x86
	at java.util.concurrent.FutureTask.report(FutureTask.java:122)
	at java.util.concurrent.FutureTask.get(FutureTask.java:192)
	at xxxx.scanEthernetIp(xxxx.java:524)
	... 4 more
Caused by: io.netty.handler.codec.DecoderException: unhandled item type: 0x86
	at com.digitalpetri.enip.cpf.CpfItem.decode(CpfItem.java:79)
	at com.digitalpetri.enip.cpf.CpfPacket.decode(CpfPacket.java:49)
	at com.digitalpetri.enip.commands.ListIdentity.decode(ListIdentity.java:54)
	at com.digitalpetri.enip.EnipPacket.decodeCommand(EnipPacket.java:163)
	at com.digitalpetri.enip.EnipPacket.decode(EnipPacket.java:127)
	at xxxx.receive(Scanner.java:65)
	at xxxx.scanNet(Scanner.java:32)
	at xxxx.scanAllNet(Scanner.java:42)
	at xxxx.scan(GlobalEthernetIpConnectionDevice.java:63)
	at xxxx.lambda$scanEthernetIp$2(xxxx.java:512)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
	... 1 more

I could see that the response contains 2 CpfItem, a CipIdentityItem and an unknown one.
Here is the response as byte array:
[99, 0, 60, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 2, 0, 12, 0, 45, 0, 1, 0, 0, 2, -81, 18, -64, -88, -56, -7, 0, 0, 0, 0, 0, 0, 0, 0, 77, 0, 12, 0, 115, 0, 11, 1, 96, 0, -23, 27, -51, 44, 11, 82, 68, 78, 69, 84, 76, 84, 69, 69, 48, 55, -1, -122, 0, 5, 0, 2, 0, 0, 0, 11, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0]

I hope someone can help me with that issue.

--- a/enip-core/src/main/java/com/digitalpetri/enip/commands/ListIdentity.java
+++ b/enip-core/src/main/java/com/digitalpetri/enip/commands/ListIdentity.java
@@ -1,8 +1,12 @@
 package com.digitalpetri.enip.commands;
 
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Objects;
 import java.util.Optional;
 
 import com.digitalpetri.enip.cpf.CipIdentityItem;
+import com.digitalpetri.enip.cpf.CipSecurityItem;
 import com.digitalpetri.enip.cpf.CpfItem;
 import com.digitalpetri.enip.cpf.CpfPacket;
 import io.netty.buffer.ByteBuf;
@@ -11,39 +15,69 @@ public final class ListIdentity extends Command {
 
     public static final CommandCode COMMAND_CODE = CommandCode.ListIdentity;
 
-    private final Optional<CipIdentityItem> identity;
+    private final CipIdentityItem identityItem;
+    private final CipSecurityItem securityItem;
 
     public ListIdentity() {
         this(null);
     }
 
-    public ListIdentity(CipIdentityItem identity) {
-        super(COMMAND_CODE);
-
-        this.identity = Optional.ofNullable(identity);
+    public ListIdentity(CipIdentityItem identityItem) {
+        this(identityItem, null);
     }
 
+    public ListIdentity(CipIdentityItem identityItem, CipSecurityItem securityItem) {
+        super(COMMAND_CODE);
+
+        this.identityItem = identityItem;
+        this.securityItem = securityItem;
+    }
+
+    /**
+     * @deprecated use {@link #getIdentityItem()}
+     */
+    @Deprecated
     public Optional<CipIdentityItem> getIdentity() {
-        return identity;
+        return Optional.ofNullable(identityItem);
+    }
+
+    /**
+     * @return an {@link Optional} containing the {@link CipIdentityItem}, if present.
+     */
+    public Optional<CipIdentityItem> getIdentityItem() {
+        return Optional.ofNullable(identityItem);
+    }
+
+    /**
+     * @return an {@link Optional} containing the {@link CipSecurityItem}, if present.
+     */
+    public Optional<CipSecurityItem> getSecurityItem() {
+        return Optional.ofNullable(securityItem);
     }
 
     @Override
     public boolean equals(Object o) {
         if (this == o) return true;
         if (o == null || getClass() != o.getClass()) return false;
-
         ListIdentity that = (ListIdentity) o;
-
-        return identity.equals(that.identity);
+        return Objects.equals(identityItem, that.identityItem) &&
+            Objects.equals(securityItem, that.securityItem);
     }
 
     @Override
     public int hashCode() {
-        return identity.hashCode();
+        return Objects.hash(identityItem, securityItem);
     }
 
     public static ByteBuf encode(ListIdentity command, ByteBuf buffer) {
-        command.getIdentity().ifPresent(identity -> CpfPacket.encode(new CpfPacket(identity), buffer));
+        List<CpfItem> items = new ArrayList<>();
+        command.getIdentityItem().ifPresent(items::add);
+        command.getSecurityItem().ifPresent(items::add);
+
+        if (items.size() > 0) {
+            CpfItem[] itemArray = items.toArray(new CpfItem[0]);
+            CpfPacket.encode(new CpfPacket(itemArray), buffer);
+        }
 
         return buffer;
     }
@@ -53,14 +87,18 @@ public final class ListIdentity extends Command {
             CpfPacket packet = CpfPacket.decode(buffer);
             CpfItem[] items = packet.getItems();
 
-            if (items.length > 0) {
-                if (items[0] instanceof CipIdentityItem) {
-                    return new ListIdentity((CipIdentityItem) items[0]);
-                } else {
-                    throw new RuntimeException(
-                        String.format("expected CipIdentityItem; received %s instead",
-                            items[0].getClass().getSimpleName()));
+            CipIdentityItem identityItem;
+            CipSecurityItem securityItem = null;
+
+            if (items.length >= 1) {
+                assert items[0] instanceof CipIdentityItem;
+                identityItem = (CipIdentityItem) items[0];
+
+                if (items.length >= 2 && items[1] instanceof CipSecurityItem) {
+                    securityItem = (CipSecurityItem) items[1];
                 }
+
+                return new ListIdentity(identityItem, securityItem);
             } else {
                 return new ListIdentity();
             }

--- a/enip-core/src/main/java/com/digitalpetri/enip/cpf/CipSecurityItem.java
+++ b/enip-core/src/main/java/com/digitalpetri/enip/cpf/CipSecurityItem.java
@@ -22,18 +22,75 @@ public class CipSecurityItem extends CpfItem {
         this.ianaPortState = ianaPortState;
     }
 
+    /**
+     * Get the CIP Security Profiles supported by the device.
+     *
+     * <pre>
+     *  Bit 0       EtherNet/IP Integrity Profile
+     *  Bit 1       EtherNet/IP Confidentiality Profile
+     *  Bit 2       CIP Authorization Profile
+     *  Bit 3       CIP Integrity Profile
+     *  Bit 4-15    Reserved
+     * </pre>
+     *
+     * @return the CIP Security Profiles supported by the device.
+     */
     public int getSecurityProfiles() {
         return securityProfiles;
     }
 
+    /**
+     * Get the current state of the CIP Security Object.
+     *
+     * <pre>
+     *  0   Factory Default Configuration
+     *  1   Initial Commissioning In Progress
+     *  2   Configured
+     *  3   Incomplete Configuration
+     * </pre>
+     *
+     * @return the current state of the CIP Security Object.
+     */
     public int getCipSecurityState() {
         return cipSecurityState;
     }
 
+    /**
+     * Get the current state of the EtherNet/IP Security Object associated with the IP address
+     * where the request was received.
+     *
+     * <pre>
+     *  0   Factory Default Configuration
+     *  1   Configuration In Progress
+     *  2   Configured
+     * </pre>
+     *
+     * @return the current state of the EtherNet/IP Security Object associated with the IP address
+     * where the request was received.
+     */
     public int getEnipSecurityState() {
         return enipSecurityState;
     }
 
+    /**
+     * Get the current state, open or closed, for all EtherNet/IP related IANA ports Object
+     * associated with the IP address where the request was received.
+     * <p>
+     * 1 (TRUE) shall indicate that the corresponding port is open. If the bit is 0 (FALSE) the
+     * port is closed. Reserved bits shall be 0.
+     *
+     * <pre>
+     *  Bit 0    44818/tcp
+     *  Bit 1    44818/udp
+     *  Bit 2    2222/udp
+     *  Bit 3    2221/tcp
+     *  Bit 4    2221/udp
+     *  Bit 5-7  Reserved
+     * </pre>
+     *
+     * @return the current state, open or closed, for all EtherNet/IP related IANA ports Object
+     * associated with the IP address where the request was received.
+     */
     public int getIanaPortState() {
         return ianaPortState;
     }

--- a/enip-core/src/main/java/com/digitalpetri/enip/cpf/CipSecurityItem.java
+++ b/enip-core/src/main/java/com/digitalpetri/enip/cpf/CipSecurityItem.java
@@ -1,0 +1,95 @@
+package com.digitalpetri.enip.cpf;
+
+import java.util.Objects;
+
+import io.netty.buffer.ByteBuf;
+
+public class CipSecurityItem extends CpfItem {
+
+    public static final int TYPE_ID = 0x86;
+
+    private final int securityProfiles;
+    private final int cipSecurityState;
+    private final int enipSecurityState;
+    private final int ianaPortState;
+
+    public CipSecurityItem(int securityProfiles, int cipSecurityState, int enipSecurityState, int ianaPortState) {
+        super(TYPE_ID);
+
+        this.securityProfiles = securityProfiles;
+        this.cipSecurityState = cipSecurityState;
+        this.enipSecurityState = enipSecurityState;
+        this.ianaPortState = ianaPortState;
+    }
+
+    public int getSecurityProfiles() {
+        return securityProfiles;
+    }
+
+    public int getCipSecurityState() {
+        return cipSecurityState;
+    }
+
+    public int getEnipSecurityState() {
+        return enipSecurityState;
+    }
+
+    public int getIanaPortState() {
+        return ianaPortState;
+    }
+
+    @Override
+    public boolean equals(Object o) {
+        if (this == o) return true;
+        if (o == null || getClass() != o.getClass()) return false;
+        CipSecurityItem that = (CipSecurityItem) o;
+        return securityProfiles == that.securityProfiles &&
+            cipSecurityState == that.cipSecurityState &&
+            enipSecurityState == that.enipSecurityState &&
+            ianaPortState == that.ianaPortState;
+    }
+
+    @Override
+    public int hashCode() {
+        return Objects.hash(securityProfiles, cipSecurityState, enipSecurityState, ianaPortState);
+    }
+
+    public static ByteBuf encode(CipSecurityItem item, ByteBuf buffer) {
+        buffer.writeShort(TYPE_ID);
+
+        // Length placeholder...
+        int lengthStartIndex = buffer.writerIndex();
+        buffer.writeShort(0);
+
+        // Encode the item...
+        int itemStartIndex = buffer.writerIndex();
+        buffer.writeShort(item.getSecurityProfiles());
+        buffer.writeByte(item.getCipSecurityState() & 0xFF);
+        buffer.writeByte(item.getEnipSecurityState() & 0xFF);
+        buffer.writeByte(item.getIanaPortState() & 0xFF);
+
+        // Go back and update the length.
+        int bytesWritten = buffer.writerIndex() - itemStartIndex;
+        buffer.markWriterIndex();
+        buffer.writerIndex(lengthStartIndex);
+        buffer.writeShort(bytesWritten);
+        buffer.resetWriterIndex();
+
+        return buffer;
+    }
+
+    public static CipSecurityItem decode(ByteBuf buffer) {
+        int typeId = buffer.readUnsignedShort();
+        assert typeId == TYPE_ID;
+
+        buffer.skipBytes(2); // length
+
+        return new CipSecurityItem(
+            buffer.readShort(),
+            buffer.readUnsignedByte(),
+            buffer.readUnsignedByte(),
+            buffer.readUnsignedByte()
+        );
+    }
+
+}

--- a/enip-core/src/main/java/com/digitalpetri/enip/cpf/CpfItem.java
+++ b/enip-core/src/main/java/com/digitalpetri/enip/cpf/CpfItem.java
@@ -40,6 +40,9 @@ public abstract class CpfItem {
             case UnconnectedDataItemRequest.TYPE_ID:
                 return UnconnectedDataItemRequest.encode((UnconnectedDataItemRequest) item, buffer);
 
+            case CipSecurityItem.TYPE_ID:
+                return CipSecurityItem.encode((CipSecurityItem) item, buffer);
+
             default:
                 throw new RuntimeException(String.format("unhandled item type: 0x%02X", item.getTypeId()));
         }
@@ -73,6 +76,9 @@ public abstract class CpfItem {
             case UnconnectedDataItemResponse.TYPE_ID:
                 return UnconnectedDataItemResponse.decode(buffer);
 
+            case CipSecurityItem.TYPE_ID:
+                return CipSecurityItem.decode(buffer);
+                
             default:
                 throw new RuntimeException(String.format("unhandled item type: 0x%02X", typeId));
         }

--- a/enip-core/src/test/java/com/digitalpetri/enip/commands/ListIdentityTest.java
+++ b/enip-core/src/test/java/com/digitalpetri/enip/commands/ListIdentityTest.java
@@ -2,6 +2,7 @@ package com.digitalpetri.enip.commands;
 
 import com.digitalpetri.enip.SerializationTest;
 import com.digitalpetri.enip.cpf.CipIdentityItem;
+import com.digitalpetri.enip.cpf.CipSecurityItem;
 import com.digitalpetri.enip.cpf.SockAddr;
 import org.testng.annotations.DataProvider;
 import org.testng.annotations.Test;
@@ -11,8 +12,8 @@ import static org.testng.Assert.assertEquals;
 public class ListIdentityTest extends SerializationTest<ListIdentity> {
 
     @Test(dataProvider = "getIdentityItem")
-    public void testSerialization(CipIdentityItem identityItem) {
-        ListIdentity identity = new ListIdentity(identityItem);
+    public void testSerialization(CipIdentityItem identityItem, CipSecurityItem securityItem) {
+        ListIdentity identity = new ListIdentity(identityItem, securityItem);
         ListIdentity decoded = encodeDecode(identity, ListIdentity::encode, ListIdentity::decode);
 
         assertEquals(identity, decoded);
@@ -24,9 +25,12 @@ public class ListIdentityTest extends SerializationTest<ListIdentity> {
             0, new SockAddr(1, 2, new byte[4], 0), 1, 2, 3,
             (short) 4, (short) 5, (short) 6, 1234L, "test", (short) 0);
 
+        CipSecurityItem securityItem = new CipSecurityItem(1, 2, 3, 4);
+
         return new Object[][]{
-            {null},
-            {identityItem}
+            {null, null},
+            {identityItem, null},
+            {identityItem, securityItem}
         };
     }
 

--- a/enip-core/src/test/java/com/digitalpetri/enip/cpf/CpfItemTest.java
+++ b/enip-core/src/test/java/com/digitalpetri/enip/cpf/CpfItemTest.java
@@ -26,6 +26,7 @@ public class CpfItemTest extends SerializationTest<CpfItem> {
             {new SequencedAddressItem(1L, 2L)},
             {new SockAddrItemO2t(new SockAddr(1, 2, new byte[]{1, 2, 3, 4}, 0L))},
             {new SockAddrItemT2o(new SockAddr(1, 2, new byte[]{1, 2, 3, 4}, 0L))},
+            {new CipSecurityItem(1, 2, 3, 4)}
         };
     }
 

